### PR TITLE
[3006.x] Don't needlessly refresh fileserver backends during Pillar rendering

### LINKED ISSUE #65990: [BUG] Fileclient requests generated as a result of Pillar SLS rendering cause fileserver backends to be refreshed
**Description**

When a pillar SLS file includes a call-out to a function from the `cp` module (either directly, or by calling a salt function which calls out to the `cp` module), this spawns a new instance of the fileclient using the current opts. When `file_client` is set to `local` (as it is during PIllar rendering), this fileclient is an instance of `salt.fileclient.FSClient`. During instantiation of this instance, the `FSClient` instantiates an instance of `salt.fileserver.FSChan`, which in turn has its own instance of the fileserver.

On a masterless salt run, this is how masterless minions interface with the fileserver backends specified in the minion config. For that reason, `salt.fileserver.FSChan` refreshes all of its fileserver backends, allowing for gitfs to be successfully used in masterless.

However, this refresh is not needed during Pillar rendering in a traditional master/minion setup, because the master has a maintenance thread which handles keeping fileserver backends up-to-date. Not only is it unneeded, it slows down Pillar rendering performing unneeded refreshes. If Pillar rendering coincides with a normal fileserver update, this slowdown is exacerbated due to the fact that the act of refreshing a gitfs remote obtains a blocking lock. As the number of `gitfs_remotes` increases, this increases the likelihood that a pillar refresh (or the initial pillar obtained when a minion connects to the master) will timeout.

**Setup**

**`/etc/salt/master`** 
```yaml

# This is just here to simplify key acceptance for the purposes of reproducing this bug
auto_accept: True

# Use gitfs only
fileserver_backend:
  - git

gitfs_remotes:
  # This repo must have a master branch containing a simple JSON file (e.g. foo.json)
  - https://github.com/someuser/somerepo.git:
    # Interval increased to an absurd length so a normal maintenance thread
    # fileserver update won't be confused for one started in pillar rendering  
    - update_interval: 1800
```

**`/srv/pillar/top.sls`**
```yaml
base:
  '*':
    - test
```

**`/srv/pillar/test.sls`**
```yaml
# The path after salt:// is the relative path to the JSON file in your gitfs repo 
foo: {{ salt.cp.get_file_str("salt://path/to/foo.json") }}
# NOTE: This is a contrived example but serves to reproduce the behavior
```

**Steps to Reproduce the behavior**

1. Bring up the master in the foreground using `salt-master -l debug`
2. Configure a minion to connect to your master's IP address.

**Expected behavior**

Wait for the minion to connect after the key has been accepted. When it does, it will request Pillar data from the master.

Because the pillar SLS file you created is making a fileclient request, there will be several lines in the master's debug logging (check the scrollback) which mention gitfs. It is normal to see some gitfs log entries here, because Salt is attaching to all the local gitfs cachedirs. However, we should _not_ see any log entries for repo locking and fetching, since those updates are handled by the maintenance thread. Nonetheless, with the above setup, you will see the following:

```
[DEBUG   ][40978] Set update lock for gitfs remote 'https://github.com/someuser/somerepo.git'
[DEBUG   ][40978] Fetching gitfs remote 'https://github.com/someuser/somerepo.git'
[DEBUG   ][40978] gitfs remote 'https://github.com/someuser/somerepo.git' is up-to-date
[DEBUG   ][40978] Removed update lock for gitfs remote 'https://github.com/someuser/somerepo.git'
```

**Versions Report**

This bug would affect all salt-master versions going back to when this fileserver update was added in fe4537f58c54a9545dbb856ec9d132820187a7a8, over 9 years ago.

### PR DESCRIPTION
**Pull request by terminalmage**
### What does this PR do?

This uses the opts key that masterless runs use to limit fileserver backend refreshes to a single refresh. Adding it to the master opts ensures that FSChan instances created on the master do not trigger fileserver backend refreshes.

### What issues does this PR fix or reference?

Fixes #65990

### Previous Behavior

Fileserver backends are refreshed when the master instantiates an `FSClient` instance.

### New Behavior

Fileserver backends are no longer refreshed.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by dmurphy18 — CHANGES_REQUESTED** [review]
The tests case needs to be under tests/pytest and using pytest to test with.
Tests under tests/unit are considered unittest and will eventually be removed.

All tests PR's need tests now written using pytest, are they will be rejected.

**Comment by terminalmage** [comment]
> The tests case needs to be under tests/pytest and using pytest to test with. Tests under tests/unit are considered unittest and will eventually be removed.
> 
> All tests PR's need tests now written using pytest, are they will be rejected.

I brought this up with @dmurphy18 in Slack, but this is directly contradicted by the following docs pages:

- https://docs.saltproject.io/en/latest/topics/development/contributing.html
- https://docs.saltproject.io/en/latest/topics/tutorials/writing_tests.html

Additionally, the location where tests should be added is _not_ covered by the recent changes to [Project Management](https://docs.saltproject.io/en/master/topics/maintenance-policy.html). Both the above two pages should arguably be linked from this page.

With two distinct sets of tests in the repo, it's a must to get this documentation updated to reflect the current policies of the project. Failure to do so increases the burden placed on the core team, as well as the contributors. 

I've moved the tests over. I'm also available to assist in the process of updating the docs so that they more accurately reflect reality, provided I can get better clarity on the changes to these policies.

File: salt/config/__init__.py
Comment (s0undt3ch) [inline salt/config/__init__.py:3961]: I don't like the fact that we are using `opts` to tweak behavior "down the pipeline".
With that said, you're not introducing that behavior, it's been long present in Salt.

**Review by s0undt3ch — DISMISSED** [review]
I updated your tests to use existing test fixtures and rebased the PR against 3006.x

**Comment by terminalmage** [comment]
You broke the tests, I'm fixing them

**Comment by terminalmage** [comment]
@s0undt3ch OK, tests are fixed now.

You can't just use the opts dict, because the way salt actually works when it stands up a master is it loads the config using `salt.config.master_config()`, passing it the path to the config file. Within that function it runs `salt.config.apply_master_config()`, passing in the defaults and anything loaded from the config file path passed to `salt.config.master_config()`.

Since we aren't loading from a config file, running `salt.config.apply_master_config()` and passing in our own overrides (and using the fixture data as the defaults) is a close enough approximation for testing purposes.


Note that I also removed the monkeypatching of the fixture dictionaries. I forgot what project the fixtures come from, but I was worried that adding a key to that dictionary would pollute other tests if minion_opts/master_opts weren't function-scoped.

File: salt/config/__init__.py
Comment (terminalmage) [reply salt/config/__init__.py:3957]: Given that fileclients are generated automatically when needed, and that only the opts are passed in (see `salt.fileclient.get_file_client()`), I don't see a better solution than to use the opts. If you have a better idea, I'm interested. However, any changes to how fileclients are spawned would have impacts to a number of other systems, which is why I didn't try to reinvent the wheel when I made my bugfix.

Comment (s0undt3ch) [reply salt/config/__init__.py:3957]: > Given that fileclients are generated automatically when needed, and that only the opts are passed in (see `salt.fileclient.get_file_client()`), I don't see a better solution than to use the opts. If you have a better idea, I'm interested. However, any changes to how fileclients are spawned would have impacts to a number of other systems, which is why I didn't try to reinvent the wheel when I made my bugfix.

That also why I did not suggest an alternate approach at this moment.

**Comment by s0undt3ch** [comment]
> You broke the tests, I'm fixing them

Hmm, funny, they did pass locally.

**Review by s0undt3ch — DISMISSED** [review]
I'm good with the latest changes

**Review by twangboy — DISMISSED** [review]

File: salt/config/__init__.py
Comment (terminalmage) [inline salt/config/__init__.py:3952]: @s0undt3ch @twangboy I know you've both already approved the PR, but this line should say FSClient instead of LocalClient. Mind if I change it before merging? Feel free to also just make this change yourself.

Comment (s0undt3ch) [reply salt/config/__init__.py:3952]: Sure, go ahead a commit the change.

**Review by twangboy — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Review by dmurphy18 — APPROVED** [review]

**Pull request merged**
